compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -2054,17 +2048,609 @@ boolean checkTypeAndVarCountConsistency(BLangTupleVariable varNode, BTupleType t
+                BLangVariable value = variable.getValue();
+                value.isDeclaredWithVar = true;
+                defineNode(value, env);
+            }
+
+            BLangSimpleVariable restParam = (BLangSimpleVariable) recordVar.restParam;
+            if (restParam != null) {
+                restParam.isDeclaredWithVar = true;
+                defineNode(restParam, env);
+            }
+            return;
+        }
+        if (recordVar.type == null) {
+            recordVar.type = symResolver.resolveTypeNode(recordVar.typeNode, env);
         }
+        // To support variable forward referencing we need to symbol enter each record member with type at SymbolEnter.
+        if (!(symbolEnterAndValidateRecordVariable(recordVar, env))) {
+            recordVar.type = symTable.semanticError;
+            return;
+        }
+    }
+
+    boolean symbolEnterAndValidateRecordVariable(BLangRecordVariable recordVar, SymbolEnv env) {
+        Name varName = names.fromString(anonymousModelHelper.getNextRecordVarKey(env.enclPkg.packageID));
+        recordVar.symbol =
+                defineVarSymbol(recordVar.pos, recordVar.flagSet, recordVar.type, varName, env, recordVar.internal);
+
+        return validateRecordVariable(recordVar, env);
+    }
+
+    boolean validateRecordVariable(BLangRecordVariable recordVar, SymbolEnv env) {
+        BRecordType recordVarType;
+        /*
+          This switch block will resolve the record type of the record variable.
+          For example consider the following -
+          type Foo record {int a, boolean b};
+          type Bar record {string a, float b};
+          Foo|Bar {a, b} = foo();
+          Since the varNode type is a union, the types of 'a' and 'b' will be resolved as follows:
+          Type of 'a' will be a union of the types of field 'a' in both Foo and Bar.
+          i.e. type of 'a' is (int | string) and type of 'b' is (boolean | float).
+          Consider anydata {a, b} = foo();
+          Here, the type of 'a'and type of 'b' will be both anydata.
+         */
+        switch (recordVar.type.tag) {
+            case TypeTags.UNION:
+                BUnionType unionType = (BUnionType) recordVar.type;
+                Set<BType> bTypes = types.expandAndGetMemberTypesRecursive(unionType);
+                List<BType> possibleTypes = bTypes.stream()
+                        .filter(rec -> doesRecordContainKeys(rec, recordVar.variableList, recordVar.restParam != null))
+                        .collect(Collectors.toList());
+
+                if (possibleTypes.isEmpty()) {
+                    dlog.error(recordVar.pos, DiagnosticErrorCode.INVALID_RECORD_BINDING_PATTERN, recordVar.type);
+                    return false;
+                }
+
+                if (possibleTypes.size() > 1) {
+                    BRecordTypeSymbol recordSymbol = Symbols.createRecordSymbol(Flags.ANONYMOUS,
```

Comment:
Can we extract the logic in this block out to a method?

Reply:
This entire method seems a bit long, can we check and refactor this?

Reply:
+1